### PR TITLE
SNOW-3784414: fix token cache key — v2 canonical-JSON SHA256 key with PascalCase prefix and lowercase normalization

### PR DESCRIPTION
**Pull request by sfc-gh-mhofman**
Introduce a versioned v2 token cache key format to fix two classes of collision:

1. **Multi-account / multi-role collisions** — OAuth flows now include `idp`, `role`, `snowflake`, and `username` in a flow-specific `keyData` map; MFA/ID flows include only `snowflake` and `username`.
2. **Case-sensitivity normalization** — all identifiers are lowercased (unquoted) or returned verbatim (any `"` present, including SQL `""` escaped quotes). URLs are lowercased after scheme and userinfo stripping.

### Key format

```
SnowflakeTokenCache.v2.<PascalCaseType>.<sha256hex(canonical_json(keyData))>
```

- `PascalCaseType`: `MfaToken`, `IdToken`, `OauthAccessToken`, `OauthRefreshToken`, `DpopBundledAccessToken`
- `keyData` fields sorted alphabetically; serialised as compact JSON before hashing.
- Applied uniformly across Windows Credential Manager and file backends.

### Golden test vectors (must not change)

- **A** (OAuth/DPoP): `SnowflakeTokenCache.v2.DpopBundledAccessToken.741b6d66d252666d6821bfd19e0151511cf4efdaaeba2b3c87673aa4de6d2c0b`
- **B** (MFA): `SnowflakeTokenCache.v2.MfaToken.10c5dde84bb8f584c0df06ea826d418c4f580e08f9db10187c0cb5e2a732a0d6`

### Changes

- `SnowflakeCredentialManagerFactory`: `CacheKeyInput` positional record, `BuildCacheKey`, `NormalizeUrl`, `NormalizeIdentifier`, `ToSha256HashLower`; removed obsolete `GetSecureCredentialKey`.
- `SFSession`: `BuildTokenCacheKey(TokenType)` helper replaces four inline `CacheKeyInput` constructions.
- `ExternalBrowserAuthenticator`: `BuildIdTokenCacheKey()` helper.
- `OAuthCacheKeys`: uses full `idp` + `role` for OAuth flows.
- `IsExternalInit.cs` shim for `readonly record struct` on net481/netstandard2.0.
- Unit tests: golden vectors A & B, dimension-isolation tests, `NormalizeUrl`/`NormalizeIdentifier` edge cases.

### Checklist
- [x] Code compiles correctly
- [x] Code is formatted according to [Coding Conventions](../blob/master/CodingConventions.md)
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing (`dotnet test`)
- [ ] Extended the README / documentation, if necessary
- [x] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,10 @@
 #### For the official .NET Release Notes please refer to https://docs.snowflake.com/en/release-notes/clients-drivers/dotnet
 
 # Changelog
+- v5.9.0
```
Comment (sfc-gh-mcepiga) [inline CHANGELOG.md:4]: 5.8.0 is not released yet, is this meant to be delayed (i.e. 5.8.0 released without it, and then this cache key changeset should be merged)?

File: Snowflake.Data/Core/Session/SFSession.cs [hunk 2]
```diff
@@ -131,13 +131,27 @@ internal void ProcessLoginResponse(LoginResponse authnResponse)
...
+                        username: properties[SFSessionProperty.USER],
+                        role: properties.TryGetValue(SFSessionProperty.ROLE, out var idTokenRole) ? idTokenRole : string.Empty
+                    ));
                     SnowflakeCredentialManagerFactory.GetCredentialManager().SaveCredentials(idTokenKey, authnResponse.data.idToken);
                 }
                 if (!string.IsNullOrEmpty(authnResponse.data.mfaToken))
                 {
                     _mfaToken = SecureStringHelper.Encode(authnResponse.data.mfaToken);
-                    var key = SnowflakeCredentialManagerFactory.GetSecureCredentialKey(properties[SFSessionProperty.HOST], properties[SFSessionProperty.USER], TokenType.MFAToken);
+                    // MFA: Snowflake is the IdP; role is not part of the MFA cache key.
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Core/Session/SFSession.cs:147]: consider adding helper like this


```
private string GetCacheKeyInput(TokenType tokenType, string overrideRole = null)
        {
            overrideRole ??= properties.TryGetValue(SFSessionProperty.ROLE, out var idTokenRole) ? idTokenRole : string.Empty;

            var cacheKeyInput = new CacheKeyInput(
                tokenType: tokenType.GetAttribute<StringAttr>().value,
                idp: properties[SFSessionProperty.HOST],
                snowflake: properties[SFSessionProperty.HOST],
                username: properties[SFSessionProperty.USER],
                role: overrideRole
            );
            return SnowflakeCredentialManagerFactory.BuildCacheKey(cacheKeyInput);
        }
```

which would simplify this, key construct lines would be transformed into

```
GetCacheKeyInput(TokenType.IdToken); 
// and

var key = GetCacheKeyInput(TokenType.MFAToken, string.Empty);
```

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 3]
```diff
@@ -1,13 +1,36 @@
...
+        public CacheKeyInput(string tokenType, string idp, string snowflake, string username, string role)
+        {
+            TokenType = tokenType;
+            Idp = idp;
+            SnowflakeUrl = snowflake;
+            Username = username;
+            Role = role;
+        }
+    }
+
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:33]:
```suggestion
internal readonly record struct CacheKeyInput(string TokenType, string Idp, string SnowflakeUrl, string Username, string Role);
```

(You'd also need to adjust named constructor arguments (lowercase => uppercase) + define this class to support older tfms: (put it in separate file in ~/PlatformShims or ~/SomeOtherAppropriatelyNamedDir)

```
namespace System.Runtime.CompilerServices;

public class IsExternalInit { }

```

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 4]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+        /// Builds a v2 token cache key: <c>SnowflakeTokenCache.v2.&lt;sha256hex(canonical_json)&gt;</c>.
+        /// All inputs are normalized before hashing.
+        /// </summary>
+        internal static string BuildCacheKey(CacheKeyInput input)
+        {
+            if (string.IsNullOrEmpty(input.SnowflakeUrl))
+                throw new ArgumentException("snowflake URL must not be empty");
+            if (string.IsNullOrEmpty(input.Username))
+                throw new ArgumentException("username must not be empty");
+            // An empty idp would silently collapse the cross-IdP dimension of the key and
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:53]: It'd cut this comment - in current cache key shape there's no reason to treat idp differently than e.g. username or snowflake url. All of them are required and that's it. This comment raises more questions than answers.

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 5]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+            var keyData = new SortedDictionary<string, string>
+            {
+                ["idp"]        = NormalizeUrl(input.Idp),
+                ["role"]       = NormalizeIdentifier(input.Role),
+                ["snowflake"]  = NormalizeUrl(input.SnowflakeUrl),
+                ["token_type"] = input.TokenType,
+                ["username"]   = NormalizeIdentifier(input.Username),
+            };
+
+            string json = JsonConvert.SerializeObject(keyData, Formatting.None);
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:67]: use compiler-inferred variable type

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 6]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+            string hash = ToSha256HashLower(json);
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:68]: use compiler-inferred variable type

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 7]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+            var slashIdx = s.IndexOf('/');
+            var authorityEnd = slashIdx >= 0 ? slashIdx : s.Length;
+            var authority = s.Substring(0, authorityEnd);
+            var path = s.Substring(authorityEnd);
+            var atIdx = authority.IndexOf('@');
+            if (atIdx >= 0)
+                authority = authority.Substring(atIdx + 1);
+
+            return (authority + path).TrimEnd('/').ToUpperInvariant();
+        }
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:100]: consider sth across the lines of

```
internal static string NormalizeUrl(string url)
        {
            if (string.IsNullOrEmpty(url))
                return string.Empty;

            var uri = new Uri(url, UriKind.RelativeOrAbsolute);
            uri = uri.IsAbsoluteUri ? uri : new Uri($"https://{uri}");
            var hostComponent = uri.IsDefaultPort && url.Contains($":{uri.Port}") ? UriComponents.HostAndPort : UriComponents.Host;
            var path = uri.GetComponents(hostComponent | UriComponents.Path, UriFormat.Unescaped);
            return path.TrimEnd('/').ToUpperInvariant();
        }
```

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 8]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+        /// <summary>
+        /// Uppercases unquoted segments (ASCII only, matching Snowflake identifier case-folding);
+        /// preserves the content of double-quoted segments verbatim (including surrounding quotes).
+        /// </summary>
+        internal static string NormalizeIdentifier(string identifier)
+        {
+            if (string.IsNullOrEmpty(identifier))
+                return string.Empty;
+            var sb = new StringBuilder(identifier.Length);
+            bool inQuotes = false;
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:111]: use compiler inferred variable type

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 9]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+            foreach (char c in identifier)
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:112]: use compiler inferred variable type

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 10]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+                authority = authority.Substring(atIdx + 1);
+
+            return (authority + path).TrimEnd('/').ToUpperInvariant();
+        }
+
+        /// <summary>
+        /// Uppercases unquoted segments (ASCII only, matching Snowflake identifier case-folding);
+        /// preserves the content of double-quoted segments verbatim (including surrounding quotes).
+        /// </summary>
+        internal static string NormalizeIdentifier(string identifier)
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:106]: can identifier contain escaped quote inside quotes?

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 11]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+            return sb.ToString();
+        }
+
+        private static string ToSha256HashLower(string text)
+        {
+            using (var sha = SHA256.Create())
+            {
+                var hash = sha.ComputeHash(Encoding.UTF8.GetBytes(text));
+                return BitConverter.ToString(hash).Replace("-", string.Empty).ToLowerInvariant();
+            }
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:127]:
```suggestion
           using var sha = SHA256.Create();
            var hash = sha.ComputeHash(Encoding.UTF8.GetBytes(text));
            return BitConverter.ToString(hash).Replace("-", string.Empty).ToLowerInvariant();
```

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs [hunk 12]
```diff
@@ -17,12 +40,99 @@ public class SnowflakeCredentialManagerFactory
...
+        }
+
+        [Obsolete("Use BuildCacheKey(CacheKeyInput) instead. This method produces a legacy key format.")]
```
Comment (sfc-gh-mcepiga) [inline Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:130]: Is there any reason to keep this?

**Review by sfc-gh-mcepiga — CHANGES_REQUESTED** [review]
haven't reviewed tests yet

File: CHANGELOG.md (hunk 1 above)
Comment (sfc-gh-mhofman) [reply CHANGELOG.md]: it's just any next version, doesnt matter which one, rebased onto master

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 7 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:94]: Using Uri wouldn't preserve port in the url which would break cross-driver contract.
Also provided way we would decode url endoded characters - again breaking the contract.
As for the uppercasing since we merged this PR with #1417 we will actually lowercase but its a minor thing.
I will add a comment to have reasoning about why this logic needs to work precisely this way

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 8 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs]: fixed

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 9 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs]: fixed

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 3 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:17]: Applied

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 11 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs]: applied

File: Snowflake.Data/Core/Session/SFSession.cs (hunk 2 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SFSession.cs]: done

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 10 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:104]: yes, sure `quote"unquote"` is a perfectly fine identifier

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 12 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs]: removed

File: Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs (hunk 7 above)
Comment (sfc-gh-mcepiga) [reply Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs:94]: > Using Uri wouldn't preserve port in the url

 it will, if we pass appropriate flag for the UriComponents. (code snippet has inverted boolean condition, it has '!' missing)

> Also provided way we would decode url endoded characters 

that's true, but didn't consider it to be a concern given we drop querystring here and host shouldn't be urlencoded
